---
title: "Pre-processing and creating analysis objects"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Pre-Processing 
* process 10x results
* format analysis objects
* This section is OPTIONAL

## Prior to running this code:
* Before starting set working directory to the **CompImmunologyWorkshopUCSF** folder
  * setwd(**"[PATH]/CompImmunologyWorkshopUCSF"**)

## Load Libraries
```{r,warning=FALSE,message=FALSE}
# Set Working Directory before starting

# Immcantation
suppressPackageStartupMessages(library(alakazam))
suppressPackageStartupMessages(library(shazam))
suppressPackageStartupMessages(library(tigger))
suppressPackageStartupMessages(library(dowser))

suppressPackageStartupMessages(library(glmGamPoi))
suppressPackageStartupMessages(library(Seurat))
suppressPackageStartupMessages(library(SeuratData))
suppressPackageStartupMessages(library(SeuratObject))
suppressPackageStartupMessages(library(Azimuth))
suppressPackageStartupMessages(library(DoubletFinder))
suppressPackageStartupMessages(library(scrapper))
suppressPackageStartupMessages(library(SingleR))

suppressPackageStartupMessages(library(reshape2))
suppressPackageStartupMessages(library(kableExtra))
suppressPackageStartupMessages(library(devtools))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(Matrix))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(gridExtra))
suppressPackageStartupMessages(library(cowplot))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(impactSingleCellToolkit))

# scRNA-Seq
dataset_loc <- "data/gex"

seed.number <- 123

# Workshop specific functions
source("resources/functions_workshop.R")
```


## Process scRNA-Seq data

* One thing to consider is using the **SoupX** R package to control for extracellular ambient RNA contamination.
  * The input for that tool are the hd5 files from cellranger per samples. The output is a is a filtered version of the barcode, features and matrix files

#### 1. Create combined gex matrix
```{r}
samples.vec   <- dir(dataset_loc)

gex.matrix <- generateCombinedMatrix(dataset_loc, samples.vec,THREADS = 1,multi.results = F,
                                     assay = "gex",min.genes.per.cell = 700,max.genes.per.cell = 2500)
```


#### 2. Create Seurat Object
```{r}
set.seed(seed.number)

# Create Seurat object
seurat.obj <- createSeuratObjectFromMatrix(
  sc.data      = gex.matrix,
  project.name = "GEX",
  npca         = 20, min.genes = 700,
  normalize = F, dim.reduction = F
)

# Add samples column
cells      <- row.names(seurat.obj@meta.data)
sample_col <- cells
sample_col <- tstrsplit(sample_col,"-")[[2]]
seurat.obj@meta.data$sample <- sample_col
```

#### 3. Identify Doublets and omit them

*  If you are running this command locally, consider reducting the number of threads.
```{r}
set.seed(seed.number)

seurat.obj <- findDoublets(seurat.obj,sample.col = "sample",threads = 15)

# Omit Doublets 
cells_to_keep  <- names(seurat.obj$doublet_finder[seurat.obj$doublet_finder %in% "Singlet"])
seurat.obj     <- seurat.obj[,colnames(seurat.obj) %in% cells_to_keep]
```

#### 4. Run SingleR for automated celltype annotations

* This annotation method performs better without per sample SCT integration

##### a. generate SingleR results and add annotations back into the Seurat Object
```{r}
set.seed(seed.number)

# Cluster all samples
seurat.temp   <- scaleAndClusterSeuratObject(seurat.obj,normalize = T,dims = 1:30,npca = 10,tsne = T)

# Generate SingleR Results
singler_results_list <- runSingleRcelltypeAnnot(seurat.temp,blueprintencode.annot = T,monaco.annot = T)

# Add celltype annotations back into Seurat Object
seurat.temp          <- addSingleRcelltypeAnnotToSeuratObject(seurat.temp,singler_results_list)
annot_meta           <- seurat.temp@meta.data[,c("main_bencode","fine_bencode","main_monaco","fine_monaco")]
annot_meta           <- annot_meta[row.names(seurat.obj@meta.data),]
seurat.obj@meta.data <- cbind(seurat.obj@meta.data, annot_meta)

rm(seurat.temp)
```

##### b. plots - SingleR
```{r}
hp1 <- plotScoreHeatmap(singler_results_list$main_bencode)
hp2 <- plotScoreHeatmap(singler_results_list$main_monaco)

hp3 <- plotScoreHeatmap(singler_results_list$fine_bencode)
hp4 <- plotScoreHeatmap(singler_results_list$fine_monaco)

hp3
hp4
```


#### 5. Omit all T cell V genes
```{r}
# Omit TCR V genes from object
all_genes  <- row.names(seurat.obj)
tcr_vgenes <- all_genes[grep("TRAV|TRBV|TRGV|TRDV",all_genes)]
seurat.obj <- seurat.obj[! row.names(seurat.obj) %in% tcr_vgenes,]
```


#### 6. Per sample SCT normalization and integration

* The Elbow plot can help determine an appropriate number of PCs to use
```{r}
set.seed(seed.number)

seurat.obj.list <- SplitObject(seurat.obj, split.by="sample")
seurat.obj.list <- lapply(seurat.obj.list, function(x) {
  SCTransform(x, return.only.var.genes = FALSE, verbose = FALSE)
})
  
features <- SelectIntegrationFeatures(object.list = seurat.obj.list, nfeatures = 2000)
  
# Set variable features for each object to avoid SCT model conflicts
for (i in 1:length(seurat.obj.list)) {
  VariableFeatures(seurat.obj.list[[i]]) <- features
}
  
# Prepare for integration
seurat.obj.list <- PrepSCTIntegration(object.list = seurat.obj.list, anchor.features = features)

# For larger cohorts it might be more ideal to select a smaller subset of the cohort
# to serve as the anchor set. In this workshop, we are using all samples.
# for example you could add in this arguement to the FindIntegrationAnchors command:
#    reference = names(seurat.obj.list)[1:2] # This would consider only 2 of the samples as anchors
anchors <- FindIntegrationAnchors(object.list = seurat.obj.list,
                                    normalization.method = "SCT", 
                                    anchor.features = features)
# Integrate data
# - integration layer will only have the variable to genes, but other layers will have all genes
# - To include all genes, features.to.integrate = all_genes # all_genes is a vector will all gene names
seurat.obj <- IntegrateData(anchorset = anchors, normalization.method = "SCT")

# Re-cluster
seurat.obj   <- scaleAndClusterSeuratObject(seurat.obj,normalize = F,dims = 1:30,npca = 10,tsne = T)
pc_elbowplot <- plotOptimalPCsforSeuratObject(seurat.obj)
```

#### 7. Azimuth Automated Cell Type annotations

* If you don't already have a reference installed, you can install the human PBMC referece 
  with the following command:
* **InstallData("pbmcref")**
```{r}
### Install reference if you don't have it. The line is current commented out
# InstallData("pbmcref")

# Change to SCT layer, which has both normalized and raw counts
DefaultAssay(seurat.obj) <- "SCT"

# Run Azimuth
results_azimuth  <- RunAzimuth(seurat.obj, reference = "pbmcref")

# Add annotations back into Seurat object
celltypes_azimuth <- results_azimuth@meta.data
annot_cols        <- names(celltypes_azimuth)
annot_cols        <- annot_cols[annot_cols %in% c("predicted.celltype.l1","predicted.celltype.l2","predicted.celltype.l3")]
celltypes_azimuth <- celltypes_azimuth[,annot_cols]

seurat.obj@meta.data <- cbind(seurat.obj@meta.data,celltypes_azimuth)
```



#### 8. QC filtering

Omit cells with a high % mitochondrial transcripts as well as cells with high expression of Platelet genes (PPBP and PF4)

##### a. omit cells with a high % mitochondrial genes
```{r}
seurat.obj <- subset(seurat.obj, subset = percent.mito < 10)
```

##### b. omit clusters dominated by platelet genes

* While preparing the workshop, 1 (sometimes 2) cluster was omitted in this section

```{r}
GENE_PANEL <- c("PPBP", "PF4")
CUTOFF     <- 0.3

# Check Platelet genes
vln_plot_pre <- VlnPlot(seurat.obj, features = GENE_PANEL, layer = "counts", log = T,ncol = 1)

# Identify clusters with high expression of platelet genes
avg_exp_per_clus <- AverageExpression(seurat.obj,features = GENE_PANEL,group.by = "seurat_clusters",assays = "SCT")
avg_exp_per_clus            <- avg_exp_per_clus$SCT %>% t() %>% as.data.frame()
row.names(avg_exp_per_clus) <- str_replace_all(row.names(avg_exp_per_clus), "g","")

avg_exp_per_clus$sum_exp_zscore <- rowSums(avg_exp_per_clus)
avg_exp_per_clus$sum_exp_zscore <- as.numeric(scale(avg_exp_per_clus$sum_exp_zscore))

# Omit clusters (there were two clusters omitted)
clusters_to_omit <- row.names(avg_exp_per_clus[avg_exp_per_clus$sum_exp_zscore > CUTOFF,])

cells_to_keep  <- names(seurat.obj$seurat_clusters[! seurat.obj$seurat_clusters %in% clusters_to_omit])
seurat.obj     <- seurat.obj[,colnames(seurat.obj) %in% cells_to_keep]

# Check Violin plots after removing clusters
vln_plot_post <- VlnPlot(seurat.obj, features = GENE_PANEL, layer = "counts", log = T,ncol = 1)
```



## SAVE Objects

### processed results
```{r}
saveRDS(seurat.obj,file = "objects/seurat.obj.processed.rds")
```

### Celltype annotation results
```{r}
# SingleR
saveRDS(singler_results_list,file = "objects/singleR_celltype_annotations.rds")

# Azimuth
saveRDS(celltypes_azimuth,file = "objects/azimuth_celltype_annotations.rds")
```


