---
title: "Filter scRNA-Seq and VDJ data"
author: "Ravi Dandekar"
date: "4/19/2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

GOAL: annotate cell types and define expanded clonotypes (B cells and T cells)

* Before starting set working directory to the CompImmunologyWorkshopUCSF folder
  * setwd("[PATH]/CompImmunologyWorkshopUCSF")

```{r,echo=FALSE,warning=FALSE}
# Set Working Directory before starting

# Immcantation
library(alakazam)
library(shazam)
library(tigger)
library(dowser)

library(Seurat)
library(dittoSeq)
library(reshape2)
library(kableExtra)
library(devtools)
library(data.table)
library(ggplot2)
library(RColorBrewer)
library(gridExtra)    # grid.arrange()
library(cowplot)
library(dplyr)
library(impactSingleCellToolkit)

# Immcantation Results
bcr_heavy_fh <- "data/results_immcantation/BCR_CSFPB_heavy_germ-pass.tsv"
bcr_light_fh <- "data/results_immcantation/BCR_CSFPB_light_germ-pass.tsv"
tcr_beta_fh  <- "data/results_immcantation/TCR_CSFPB_heavy_germ-pass.tsv"
tcr_alpha_fh <- "data/results_immcantation/TCR_CSFPB_light_germ-pass.tsv"

source("resources/functions_workshop.R")
```


# 1. Load RNA-Seq object

At this stage of the analysis scRNA-Seq, BCR and TCR repertoire datasets have been processed and overlapped
* Doublet were removed computationally using 'DoubletFinder'
* cells with more than 10% mitochondrial gene expression were omitted
* Only B cells and T cells exist in this object

```{r}
rnaseq_obj <- readRDS("objects/seurat.obj.processed.rds")
```

# 2. Load and format TCR/BCR repertoire results

* Clonotype definitions for BCR are based on the Heavy chain only
* Clonotype definitions for TCR include both Alpha and Beta chains

### a. Load immcantations results
```{r}
# TCR
db.tra   <- readChangeoDb(tcr_alpha_fh) %>% as.data.frame()
db.trb   <- readChangeoDb(tcr_beta_fh) %>% as.data.frame()
# BCR
db.heavy <- readChangeoDb(bcr_heavy_fh) %>% as.data.frame()
db.light <- readChangeoDb(bcr_light_fh) %>% as.data.frame()
```

### b. filter contigs and merge all chains
* All contigs are in frame
```{r}
bcr_df <- mergeVDJresults(db.heavy,db.light,umi.thresh = 3,assay = "bcr")
tcr_df <- mergeVDJresults(db.trb,db.tra,umi.thresh = 2, assay = "tcr")

# format VDJ unique cell IDs to match scRNA-Seq (function only for this workshop)
formatUniqueIDcolum <- function(uni_id_column,sep = ":") {
  samples <- tstrsplit(uni_id_column, sep)[[1]]
  cells   <- tstrsplit(uni_id_column, sep)[[2]]
  fmt_ids <- paste(cells,samples,sep = "-")
  
  return(fmt_ids)
}

bcr_df$unique_cell_id <- formatUniqueIDcolum(bcr_df$unique_cell_id)
tcr_df$unique_cell_id <- formatUniqueIDcolum(tcr_df$unique_cell_id)
```


# 3. Only analyze cells shared across all datasets

### a. get vector of overlapping cells
```{r}
cells_rnaseq      <- colnames(rnaseq_obj)
cells_vdj         <- unique(c(bcr_df$unique_cell_id,
                              tcr_df$unique_cell_id))
overlapping_cells <- cells_rnaseq[cells_rnaseq %in% cells_vdj]
```

### b. filter rnaseq and vdj data
```{r,fig.height=6,fig.width=11}
# generate UMAP of cell type annotations pre-filtering
p1_pre_filter <- UMAPPlot(object = rnaseq_obj, group.by="main_bencode", pt.size=0.3, label = T)

# only keep cells with vdj
rnaseq_obj     <- rnaseq_obj[,colnames(rnaseq_obj) %in% overlapping_cells]
bcr_df         <- bcr_df[bcr_df$unique_cell_id %in% overlapping_cells,]
tcr_df         <- tcr_df[tcr_df$unique_cell_id %in% overlapping_cells,]

# post filter umap
p2_post_filter <- UMAPPlot(object = rnaseq_obj, group.by="main_bencode", pt.size=0.3, label = T)
plot_grid(p1_pre_filter,p2_post_filter,ncol = 2)
```

# 4. Re-cluster
```{r,fig.height=6,fig.width=7}
rnaseq_obj <- scaleAndClusterSeuratObject(rnaseq_obj,dims = 1:5,npca = 5,resolution = 0.3,
                                          normalize = F,dim.reduction = T)

p3_reclustered <- UMAPPlot(object = rnaseq_obj, group.by="main_bencode", pt.size=0.3, label = T)
```


# SAVE Processed objects
```{r}
# 2022 command
#save(rnaseq_obj,tcell_obj,bcell_obj,bcr_results,tcr_results,clone_counts_tcr,clone_counts_bcr,file = "objects/processed_objects.RData")
save(rnaseq_obj,bcr_df,tcr_df,file = "objects/processed_objects.RData")
```
