---
title: "Identify mAb/mTCR Candidates and Perform Network Analysis"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Prior to running this code:
* Before starting set working directory to the CompImmunologyWorkshopUCSF folder
  * setwd("[PATH]/CompImmunologyWorkshopUCSF")


```{r,warning=FALSE,message=FALSE}
# Set Working Directory before starting

# Immcantation
suppressPackageStartupMessages(library(alakazam))
suppressPackageStartupMessages(library(shazam))
suppressPackageStartupMessages(library(tigger))
suppressPackageStartupMessages(library(dowser))

suppressPackageStartupMessages(library(glmGamPoi))
suppressPackageStartupMessages(library(Seurat))
suppressPackageStartupMessages(library(dittoSeq))
suppressPackageStartupMessages(library(reshape2))
suppressPackageStartupMessages(library(kableExtra))
suppressPackageStartupMessages(library(devtools))
suppressPackageStartupMessages(library(data.table))
suppressPackageStartupMessages(library(gplots))
suppressPackageStartupMessages(library(pheatmap))
suppressPackageStartupMessages(library(ggplot2))
suppressPackageStartupMessages(library(RColorBrewer))
suppressPackageStartupMessages(library(gridExtra))
suppressPackageStartupMessages(library(cowplot))
suppressPackageStartupMessages(library(dplyr))
suppressPackageStartupMessages(library(impactSingleCellToolkit))

seed.number <- 1234

# GLIPH2 results directory
results_dir_gliph2 <- "4_FindCandidates/all_CSF_GLIPH2"
dir.create(results_dir_gliph2,recursive = T)

# Bonus section inputs
counts_pep_fh  <- "data/input_antigen_discovery/input_counts_mAbs_phipseq_peptidome.csv"
counts_vir_fh  <- "data/input_antigen_discovery/input_counts_mAbs_phipseq_virome.csv"
results_pep_fh <- "data/input_antigen_discovery/results_enriched_peptides_peptidome.csv"
results_vir_fh <- "data/input_antigen_discovery/results_enriched_peptides_virome.csv"


source("resources/functions_workshop.R")
```


# Load Processed Objects
```{r,warning=FALSE}
load("objects/processed_objects.RData")
```

### -----------------------------

# Immune Repertoires

## --function to format networks
```{r}
# formatIgraphNetwork
#  - format networks for node and edge features so everything is consistent
formatIgraphNetwork <- function(net,extra.size = TRUE,edge.width = 3) {
  # Generate colors based on media type:
  color_vec <- V(net)$compartment
  if("PBMC" %in% color_vec){
    color_vec[color_vec %in% "PBMC"] <- "tomato"
  }
  if("CSF" %in% color_vec){
    color_vec[color_vec %in% "CSF"]  <- "darkblue"
  }
  

  V(net)$color <- color_vec
  
  # Set node size based on audience size:
  size_vec <- log10(as.numeric(V(net)$clone_size)) + 5
  if(extra.size){
    size_vec[size_vec > 5] <- size_vec[size_vec > 5] + 5
  }
  V(net)$size <- size_vec

  # Setting them to NA will render no labels:
  V(net)$label <- NA

  #change arrow size and edge color:
  E(net)$arrow.mode <- 0
  E(net)$edge.color <- "grey"
  E(net)$width      <- edge.width
  
  return(net)
}

```

# 1. BCR networks

### a. Disease patient
```{r}
bcr_df$patient <- tstrsplit(bcr_df$sample,"_")[[1]]

# CSF and PB
bcr_node_df             <- createRepertoireNodeDataframe(bcr_df,assay = "bcr",select.patients = "28",
                                                         expansion.thresh = 2)
bcr_node_df$compartment <- tstrsplit(bcr_node_df$id,"_")[[2]]
bcr_edges               <- createRepertoireEdgeDataframe(bcr_node_df)
bcr_net                 <- graph_from_data_frame(d=bcr_edges, vertices=bcr_node_df, directed=F)
bcr_net                 <- formatIgraphNetwork(bcr_net)
```

### b. Healthy patient
```{r}
# CSF and PB
bcr_hc_node_df             <- createRepertoireNodeDataframe(bcr_df,assay = "bcr",select.patients = "32",
                                                            expansion.thresh = 2)
bcr_hc_node_df$compartment <- tstrsplit(bcr_hc_node_df$id,"_")[[2]]
bcr_hc_edges               <- createRepertoireEdgeDataframe(bcr_hc_node_df)
bcr_hc_net                 <- graph_from_data_frame(d=bcr_hc_edges, vertices=bcr_hc_node_df, directed=F)
bcr_hc_net                 <- formatIgraphNetwork(bcr_hc_net)
```

### c. Plot
* Some different layouts:
  * layout_nicely
  * layout_components
  * layout_with_kk
  * layout_with_lgl
```{r}
set.seed(seed.number)

par(mfrow=c(1,2))
plot(bcr_net,layout = layout_with_kk, main = "BCR - Disease Patient")
plot(bcr_hc_net,layout = layout_with_kk, main = "BCR - Healthy Patient")
```

### d. Write GML file
```{r}
bcr_full_node_df             <- createRepertoireNodeDataframe(bcr_df,assay = "bcr",
                                                              select.patients = c("32","28"),
                                                              expansion.thresh = 2)
bcr_full_node_df$compartment <- tstrsplit(bcr_full_node_df$id,"_")[[2]]
bcr_full_edges               <- createRepertoireEdgeDataframe(bcr_full_node_df)
bcr_full_net                 <- graph_from_data_frame(d=bcr_full_edges, vertices=bcr_full_node_df, directed=F)
bcr_full_net                 <- formatIgraphNetwork(bcr_full_net)

bcr_graph_all<- graph.data.frame(bcr_full_edges, directed=F,vertices=bcr_full_node_df)
```

```{r,eval=FALSE}
write.graph(bcr_graph_all, file = "4_FindCandidates/bcell_network.gml", format = "gml")
```

# 2. TCR networks - Expanded only

T cell networks are a lot larger than B cell networks, therefore we will focus only on clonally expanded cells.

### a. Disease patient
```{r}
tcr_df$patient <- tstrsplit(tcr_df$sample,"_")[[1]]
tcr_df_exp <- tcr_df[! tcr_df$EXPANSION_CATEGORY %in% "Non-expanded",]

# CSF and PB
tcr_node_df             <- createRepertoireNodeDataframe(tcr_df_exp,assay = "tcr",select.patients = "28",
                                                         expansion.thresh = 2)
tcr_node_df$compartment <- tstrsplit(tcr_node_df$id,"_")[[2]]
tcr_edges               <- createRepertoireEdgeDataframe(tcr_node_df)
tcr_net                 <- graph_from_data_frame(d=tcr_edges, vertices=tcr_node_df, directed=F)
tcr_net                 <- formatIgraphNetwork(tcr_net,extra.size = F,edge.width = 1)
```

### b. Healthy patient
```{r}
# CSF and PB
tcr_hc_node_df             <- createRepertoireNodeDataframe(tcr_df_exp,assay = "tcr",select.patients = "32",
                                                            expansion.thresh = 2)
tcr_hc_node_df$compartment <- tstrsplit(tcr_hc_node_df$id,"_")[[2]]
tcr_hc_edges               <- createRepertoireEdgeDataframe(tcr_hc_node_df)
tcr_hc_net                 <- graph_from_data_frame(d=tcr_hc_edges, vertices=tcr_hc_node_df, directed=F)
tcr_hc_net                 <- formatIgraphNetwork(tcr_hc_net,extra.size = F,edge.width = 1)
```


### c. Plot
* Some different layouts:
  * layout_nicely
  * layout_components
  * layout_with_kk
  * layout_with_lgl
  * layout_with_fr
```{r}
set.seed(seed.number)

par(mfrow=c(1,2))
plot(tcr_net,layout = layout_with_fr, main = "TCR - Disease Patient")
plot(tcr_hc_net,layout = layout_with_fr, main = "TCR - Healthy Patient")
```

### d. Write GML file
```{r}
tcr_exp_node_df             <- createRepertoireNodeDataframe(tcr_df_exp,assay = "tcr",
                                                              select.patients = c("28","32"),
                                                              expansion.thresh = 2)
tcr_exp_node_df$compartment <- tstrsplit(tcr_exp_node_df$id,"_")[[2]]
tcr_exp_edges               <- createRepertoireEdgeDataframe(tcr_exp_node_df)
tcr_exp_net                 <- graph_from_data_frame(d=tcr_exp_edges, vertices=tcr_exp_node_df, directed=F)
tcr_exp_net                 <- formatIgraphNetwork(tcr_exp_net,extra.size = F,edge.width = 1)

tcr_graph_exp<- graph.data.frame(tcr_exp_edges, directed=F,vertices=tcr_exp_node_df)
```

```{r,eval=FALSE}
write.graph(tcr_graph_exp, file = "4_FindCandidates/tcell_expanded_network.gml", format = "gml")
```


# 3. TCR networks - CD8 Expanded only

### a. Disease patient
```{r}
tcr_df_exp <- tcr_df[! tcr_df$EXPANSION_CATEGORY %in% "Non-expanded",]
tcr_df_exp <- tcr_df_exp[tcr_df_exp$SUBTYPE_ANNOTATION %in% "CD8 T cells",]

# CSF and PB
tcr_node_df             <- createRepertoireNodeDataframe(tcr_df_exp,assay = "tcr",select.patients = "28",
                                                         expansion.thresh = 2)
tcr_node_df$compartment <- tstrsplit(tcr_node_df$id,"_")[[2]]
tcr_edges               <- createRepertoireEdgeDataframe(tcr_node_df)
tcr_net                 <- graph_from_data_frame(d=tcr_edges, vertices=tcr_node_df, directed=F)
tcr_net                 <- formatIgraphNetwork(tcr_net,extra.size = F,edge.width = 1)
```

### b. Healthy patient
```{r}
# CSF and PB
tcr_hc_node_df             <- createRepertoireNodeDataframe(tcr_df_exp,assay = "tcr",select.patients = "32",
                                                            expansion.thresh = 2)
tcr_hc_node_df$compartment <- tstrsplit(tcr_hc_node_df$id,"_")[[2]]
tcr_hc_edges               <- createRepertoireEdgeDataframe(tcr_hc_node_df)
tcr_hc_net                 <- graph_from_data_frame(d=tcr_hc_edges, vertices=tcr_hc_node_df, directed=F)
tcr_hc_net                 <- formatIgraphNetwork(tcr_hc_net,extra.size = F,edge.width = 1)
```


### c. Plot
```{r}
set.seed(seed.number)

par(mfrow=c(1,2))
plot(tcr_net,layout = layout_with_fr, main = "TCR CD8 - Disease Patient")
plot(tcr_hc_net,layout = layout_with_fr, main = "TCR CD8 - Healthy Patient")
```



# 4. Dot plot of clonotypes

* Search for Disease Relevant Clonotypes (Thresholds based on prior analysis)
    * CSF_PB_RATIO_THRESH > 1.5
    * PB_count <= 1

```{r}
clone_counts_tcr_disease <- createCloneCountTable(tcr_df[tcr_df$patient %in% "28",])
clone_counts_bcr_disease <- createCloneCountTable(bcr_df[bcr_df$patient %in% "28",])

# Add column of CSF expanded/Other
CSF_PB_RATIO_THRESH = 1.5

# TCR
csf_high <- clone_counts_tcr_disease[clone_counts_tcr_disease$PB_count <= 1 & clone_counts_tcr_disease$CSF_PB_ratio >= CSF_PB_RATIO_THRESH,]
csf_high_clonotypes <- row.names(csf_high)

exp_category <- row.names(clone_counts_tcr_disease)
exp_category[exp_category %in% csf_high_clonotypes]   <- "CSF_high_expanded"
exp_category[! exp_category %in% "CSF_high_expanded"] <- "Other"
clone_counts_tcr_disease$exp_category <- exp_category

# BCR
csf_high <- clone_counts_bcr_disease[clone_counts_bcr_disease$PB_count <= 1 & clone_counts_bcr_disease$CSF_PB_ratio >= CSF_PB_RATIO_THRESH,]
csf_high_clonotypes <- row.names(csf_high)

exp_category <- row.names(clone_counts_bcr_disease)
exp_category[exp_category %in% csf_high_clonotypes]   <- "CSF_high_expanded"
exp_category[! exp_category %in% "CSF_high_expanded"] <- "Other"
clone_counts_bcr_disease$exp_category <- exp_category
```

* Most B cell clonotypes are 1 cell either in the CSF or PB
```{r,fig.height=3, fig.width=11,warning=FALSE}
# fig.width=11 | fig.width=10 for workshop

# TCR 
p1 <- ggplot(clone_counts_tcr_disease, aes(x=pct_PB, y=pct_CSF,color=exp_category)) + geom_point(aes(size=cell_count)) + scale_color_manual(breaks = c("CSF_high_expanded", "Other"),values=c("red", "black")) + ggtitle("T cells")

p2 <- ggplot(clone_counts_bcr_disease, aes(x=pct_PB, y=pct_CSF,color=exp_category)) + geom_point(aes(size=cell_count)) + scale_color_manual(breaks = c("CSF_high_expanded", "Other"),values=c("red", "black")) + ggtitle("B cells")

#p1 <- p1 + scale_x_continuous(trans = 'log2') + scale_y_continuous(trans = 'log2')
grid.arrange(p1,p2, nrow = 1)
```

### -----------------------------

# GLIPH2 Network Analysis

## --function to format networks
```{r}
# format TCR GLIPH network
#  - interactions are based on the GLIPH group, not clone ID
formatIgraphNetworkGLIPH <- function(net,extra.size = TRUE,edge.width = 3,vertex.label.cex = 1) {
  # Generate colors based on media type:
  color_vec <- V(net)$patient
  if("32" %in% color_vec){
    color_vec[color_vec %in% "32"] <- "tomato"
  }
  if("28" %in% color_vec){
    color_vec[color_vec %in% "28"]  <- "darkblue"
  }
  

  V(net)$color <- color_vec
  
  # Set node size based on audience size:
  size_vec <- log10(as.numeric(V(net)$clone_size)) + 5
  if(extra.size){
    size_vec[size_vec > 5] <- size_vec[size_vec > 5] + 5
  }
  V(net)$size <- size_vec

  # Setting them to NA will render no labels:
  V(net)$label       <- V(net)$net_node_label
  V(net)$label.cex   <- vertex.label.cex
  V(net)$label.color <- "black"
  V(net)$label.font  <- 2
  

  #change arrow size and edge color:
  E(net)$arrow.mode <- 0
  E(net)$edge.color <- "black"
  E(net)$width      <- edge.width
  
  return(net)
}
```



# 1. Generate input files
* Running GLIPH2 is optional, results files provided
* generateGLIPH2inputFiles will create all the input files you need for GLIPH2, but running GLIPH2 still requires a parameter plain text file (made manually)

```{r}
tcr_df_csf         <- tcr_df[tcr_df$COMPARTMENT %in% "CSF",]
tcr_df_csf$patient <- tstrsplit(tcr_df_csf$sample,"_")[[1]]

generateGLIPH2inputFiles(tcr_df_csf,results.dir = results_dir_gliph2)
```


# 2. Run GLIPH2 in the terminal (OPTIONAL)

### a. Format input parameter file
* GLIPH2 input files have been provided as part of the workshop
* Input parameter file: 4_FindCandidates/all_CSF_GLIPH2/param_file_gliph2.cfg
* Example: parameter file was manually created with the following lines:
```{r,eval=FALSE}
out_prefix=AllCSF
cdr3_file=TCR_input_table.GLIPH2.tsv
refer_file=ref.txt
v_usage_freq_file=ref_V.txt
cdr3_length_freq_file=ref_L.txt
local_min_pvalue=0.001
p_depth=1000
global_convergence_cutoff=1
simulation_depth=1000
kmer_min_depth=3
local_min_OVE=10
algorithm=GLIPH2
all_aa_interchangeable=1
```

* To inlcude HLA genotypes, create a TSV with all class 1 and 2 alleles (1 row per patient)
* Example: adding HLA geneotypes in parameter file (param_file_gliph2.cfg):
```{r, eval=FALSE}
hla_file=HLA_table_formatted_GLIPH.class1and2.tab
```

### b. Run GLIPH2 in terminal (OPTIONAL)
* Results provided used input patient HLA genotypes
```{r,eval=FALSE}
cd 4_FindCandidates/all_CSF_GLIPH2
irtools.centos -c param_file_gliph2.cfg
```


# 3. Load GLIPH2 results for all CSF clones
```{r}
gliph_results_fh <- "4_FindCandidates/all_CSF_GLIPH2/AllCSF_cluster.csv"
gliph_resuls_csf <- read.csv(gliph_results_fh,header = T,stringsAsFactors = F)
gliph_resuls_csf$CRD3ab_AA <- paste(gliph_resuls_csf$TcRa,gliph_resuls_csf$TcRb,sep = ":")
all_groups       <- unique(gliph_resuls_csf$index)
```

# 2. Add CSF clonotype frequencies and cell counts to TCR results
```{r}
tcr_results_gliph <- addCloneCountColToVDJresults(results = tcr_df,clone_counts = clone_counts_tcr,target.col = "pct_CSF",new.col.name = "clonotype_CSF_freq")
tcr_results_gliph <- addCloneCountColToVDJresults(results = tcr_results_gliph,clone_counts = clone_counts_tcr,target.col = "cell_count",new.col.name = "cell_count")
tcr_results_gliph <- addCloneCountColToVDJresults(results = tcr_results_gliph,clone_counts = clone_counts_tcr,target.col = "CSF_count",new.col.name = "CSF_count")
```

# 3. Add GLIPH results into TCR table

### a. GLIPH status 
```{r}
# Add Unified CDR3 Alpha/Beta string to TCR table
tcr_cdr3_col <- tcr_results_gliph$unique_cell_id

for (cell in unique(tcr_results_gliph$unique_cell_id)) {
  cell_df <- tcr_results_gliph[tcr_results_gliph$unique_cell_id %in% cell,]
  alpha   <- cell_df[cell_df$locus %in% "TRA",]
  beta    <- cell_df[cell_df$locus %in% "TRB",]
  ab      <- paste(c(alpha$junction_aa,beta$junction_aa),collapse = ":")
  tcr_cdr3_col[tcr_cdr3_col %in% cell] <- ab
}
tcr_results_gliph$CDR3ab_AA <- tcr_cdr3_col

# Gliph results status 
gliph_status_col <- tcr_results_gliph$CDR3ab_AA
gliph_status_col[gliph_status_col %in% unique(gliph_resuls_csf$CRD3ab_AA)] <- "1"
gliph_status_col[! gliph_status_col %in% "1"] <- "0"
tcr_results_gliph$gliph_status <- gliph_status_col
```

### b. Add index groups (GLIPH interaction groups)
```{r}
target_cells <- tcr_results_gliph[tcr_results_gliph$CDR3ab_AA %in% gliph_resuls_csf$CRD3ab_AA,]
target_cells <- unique(target_cells$unique_cell_id)

index_col        <- tcr_results_gliph$CDR3ab_AA
names(index_col) <- tcr_results_gliph$unique_cell_id
index_col[! names(index_col) %in% target_cells] <- ""

target_cdr3ab <- unique(as.character(index_col[! index_col %in% ""]))


for (ab in target_cdr3ab) {
  gliph_sub <- gliph_resuls_csf[gliph_resuls_csf$CRD3ab_AA %in% ab,]
  if(nrow(gliph_sub) > 0){
    index_vec <- unique(gliph_sub$index)
    if(length(index_vec) > 1){
      index_fmt <- paste(sort(index_vec),collapse = ":")
      index_col[index_col %in% ab] <- index_fmt
    } else{
      index_col[index_col %in% ab] <- index_vec
    }
    
  }
}

tcr_results_gliph$gliph_group <- index_col
```


# 4. Filter TCR and GLIPH data based on highly expanded CSF clones
```{r}
THRESH <- 0.75  # CSF:PB ratio threshold for TCR

clonotype_counts_csf     <- clone_counts_tcr[clone_counts_tcr$CSF_count > 1,]
clonotype_counts_csf_exp <- clonotype_counts_csf[clonotype_counts_csf$cell_count > 1,]
high_csf_clones          <- clonotype_counts_csf_exp[clonotype_counts_csf_exp$CSF_PB_ratio > THRESH,]
moderate_csf_clones      <- clonotype_counts_csf_exp[clonotype_counts_csf_exp$CSF_PB_ratio <= THRESH,]


target_clones <- row.names(high_csf_clones) # Expanded CSF clones

# Omit cells with no Alpha CDR3 (no incomplete TCRs)
data_tcr_na     <- tcr_results_gliph[is.na(tcr_results_gliph$junction_aa),]
incomplete_tcrs <- unique(data_tcr_na$unique_cell_id)
data_tcr_clean  <- tcr_results_gliph[! tcr_results_gliph$unique_cell_id %in% incomplete_tcrs,]

# Get target CSF high Clonotypes
data_tcr_csf     <- data_tcr_clean[data_tcr_clean$COMPARTMENT %in% c("CSF"),]
data_tcr_csf_exp <- data_tcr_csf[data_tcr_csf$clone_id %in% row.names(clonotype_counts_csf_exp),]
data_tcr_csf_exp <- data_tcr_csf_exp[data_tcr_csf_exp$gliph_status %in% "1",]

data_tcr_target <- data_tcr_csf_exp[data_tcr_csf_exp$clone_id %in% target_clones,]

# Isoloate all cells with interacting CDR3s (all GLIPH 2 groups)
target_groups <- unique(data_tcr_target$gliph_group)

target_groups_all <- c()
for (grp in target_groups) {
  if(grepl(":",grp)){
    # string contains more than one TCR group
    # Add in full grp interaction string
    target_groups_all <- c(target_groups_all,grp)
    
    # Add in all individual groups within interaction string
    grp_list <- unlist(strsplit(grp, ":"))
    target_groups_all <- c(target_groups_all,grp_list)
  } else{
    target_groups_all <- c(target_groups_all,grp)
  }
}
target_groups_all <- unique(target_groups_all)

interac_tcr <- data_tcr_csf[data_tcr_csf$gliph_group %in% target_groups_all,]
interac_tcr <- unique(interac_tcr$unique_cell_id)

# Get all CSF high and cells with interacting TCRs
all_target_cells <- c(unique(data_tcr_target$unique_cell_id),interac_tcr)
data_tcr_target_all <- data_tcr_clean[data_tcr_clean$unique_cell_id %in% all_target_cells,]


# Omit nodes with no interactions (gliph  interactions = 0)
data_tcr_target_all <- data_tcr_target_all[data_tcr_target_all$gliph_status %in% "1",]
```

# 5. Create TCR network based on GLIPH interaction groups

### a. create network
```{r}
# Create Node dataframe
tcr_gliph_node_df             <- createRepertoireNodeDataframe(data_tcr_target_all,assay = "tcr",
                                                         clone.col = "gliph_group",
                                                         prefix.clone.id = "gliph",
                                                         expansion.thresh = 2)
# add back in some info such as clonotype ID
row.names(tcr_gliph_node_df) <- tcr_gliph_node_df$id

extra_df <- data_tcr_target_all[data_tcr_target_all$locus %in% "TRB",]
row.names(extra_df) <- extra_df$unique_cell_id
extra_df            <- extra_df[,c("clone_id","patient","COMPARTMENT")]
names(extra_df)[1]  <- "clone_id_vdj"

tcr_gliph_node_df <- merge(tcr_gliph_node_df,extra_df,by=0)
tcr_gliph_node_df[,1] <- NULL

# Create labels for the network figure
net_label_col <- tcr_gliph_node_df$clone_id
for (grp in unique(net_label_col)) {
  grp_sub <- net_label_col[net_label_col %in% grp]
  grp_sub[2:length(grp_sub)] <- ""
  net_label_col[net_label_col %in% grp] <- grp_sub
}
tcr_gliph_node_df$net_node_label <- net_label_col

# Create Edge data frame
tcr_gliph_edges         <- createRepertoireEdgeDataframe(tcr_gliph_node_df)

# Create Network object
tcr_gliph_net           <- graph_from_data_frame(d=tcr_gliph_edges, vertices=tcr_gliph_node_df, directed=F)
tcr_gliph_net           <- formatIgraphNetworkGLIPH(tcr_gliph_net,extra.size = F,edge.width = 2,vertex.label.cex = 0.9)
```

### b. plot
* red = healthy patient
* blue = disease patient
```{r}
set.seed(seed.number)

plot(tcr_gliph_net,layout = layout_with_fr, main = "TCR GLIPH Net - CSF highly expanded")
```

### -----------------------------


# Format Results for a Query in VDJdb (CSF only)

format CSF TCR results for upload to VDJ db

# 1. Add clonotype frequencies to TCR CSF results
```{r}
# Transfer clonotype Freq back to full results
tcr_results_csf <- tcr_results_gliph[tcr_results_gliph$COMPARTMENT %in% "CSF",]
tcr_results_csf <- tcr_results_csf[tcr_results_csf$patient %in% "28",]
clonotype_freq    <- tcr_results_csf$clone_id
clonotype_count   <- tcr_results_csf$clone_id

for(clone in row.names(clone_counts_tcr)){
  clone_row  <- clone_counts_tcr[row.names(clone_counts_tcr) %in% clone,]
  cell_count <- clone_row$CSF_count
  clone_freq <- clone_row$pct_CSF / 100
  
  clonotype_count[clonotype_count %in% clone] <- cell_count
  clonotype_freq[clonotype_freq %in% clone]   <- clone_freq
}

tcr_results_csf$CLONE_COUNT <- clonotype_count
tcr_results_csf$CLONE_FREQ  <- clonotype_freq
```

# 2. format table to meet VDJdb format
```{r}
cols_to_keep  <- c("CLONE_COUNT","CLONE_FREQ","junction","junction_aa","v_call","d_call","j_call")
fmt_col_names <- c("count","frequency","CDR3nt","CDR3aa","V","D","J")

fmt_tcr_results <- tcr_results_csf[,cols_to_keep]
names(fmt_tcr_results) <- fmt_col_names

fmt_tcr_results$D      <- "."
fmt_tcr_results$Vend   <- ""
fmt_tcr_results$Dstart <- ""
fmt_tcr_results$Dend   <- ""
fmt_tcr_results$Jstart <- ""

```

# 3. Output table and upload to VDJdb

VDJdb <https://vdjdb.cdr3.net/>
1. You will need to create an account or login temporarily.
2. Once logged in, click on the 'Annotation' tab
3. Click on the 'Upload' button
4. Click 'Add' and select the file we are generating bellow 
```{r,eval=FALSE}
write.table(fmt_tcr_results,file = "4_FindCandidates/table_for_VDJdb_CSF.txt",quote = F,row.names = F,sep = "\t")
```


### -----------------------------

# Assembling Monoclonals

# 1. BCR - assemble FR and CDRs for clonotypes of interest
```{r}
# B cells
target_bcr <- clone_counts_bcr[clone_counts_bcr$CSF_count >= 2,]
target_bcr <- row.names(target_bcr)[1]

# subset all cells in target clone (Heavy and Light chains)
target_bcr_df <- bcr_df[bcr_df$clone_id %in% target_bcr,] # heavy chains only

# Subset columns
cols_to_keep <- c("unique_cell_id","locus","v_call","d_call","j_call","clone_id",
                  "fwr1","cdr1","fwr2","cdr2","fwr3","cdr3","fwr4")
target_bcr_df <- target_bcr_df[,cols_to_keep]

# concatenate sequences and remove gaps
target_bcr_df$antibody_seq <- paste(target_bcr_df$fwr1,target_bcr_df$cdr1,
                                      target_bcr_df$fwr2,target_bcr_df$cdr2,
                                      target_bcr_df$fwr3,target_bcr_df$cdr3,target_bcr_df$fwr4,sep = "")
target_bcr_df$antibody_seq <- str_replace_all(target_bcr_df$antibody_seq,"\\.","")

target_bcr_df_sub <- target_bcr_df[,c("unique_cell_id","locus","v_call","d_call","j_call","clone_id","antibody_seq")]
```

# 2. TCR - assemble FR and CDRs for clonotypes of interest
```{r}
# T cells
target_tcr <- clone_counts_tcr[clone_counts_tcr$CSF_PB_ratio > 1,]
target_tcr <- target_tcr[order(target_tcr$CSF_PB_ratio,decreasing = T),]
target_tcr <- row.names(target_tcr)[1]

# subset all cells in target clone (Alpha and Beta chains)
target_tcr_df <- tcr_df[tcr_df$clone_id %in% target_tcr,] # Alpha and Beta

# Subset columns
cols_to_keep <- c("unique_cell_id","locus","v_call","d_call","j_call","clone_id",
                  "fwr1","cdr1","fwr2","cdr2","fwr3","cdr3","fwr4")
target_tcr_df <- target_tcr_df[,cols_to_keep]

# concatenate sequences and remove gaps
target_tcr_df$tcr_seq <- paste(target_tcr_df$fwr1,target_tcr_df$cdr1,
                               target_tcr_df$fwr2,target_tcr_df$cdr2,
                               target_tcr_df$fwr3,target_tcr_df$cdr3,target_tcr_df$fwr4,sep = "")
target_tcr_df$tcr_seq <- str_replace_all(target_tcr_df$tcr_seq,"\\.","")

target_tcr_df_sub <- target_tcr_df[,c("unique_cell_id","locus","v_call","d_call","j_call","clone_id","tcr_seq")]
```

# 3. Other tools

### --stichr (TCRs only)
* https://github.com/JamieHeather/stitchr
* Documentation: https://jamieheather.github.io/stitchr/
* Only makes used of TCR Beta chains
* This tool must be used in the terminal to access the GUI
* Example Entry into GUI (V/J/CDR3 AA of top expanded TCR in the CSF)
  * Beta Chain
    * TRBV7-9 
    * TRBJ1-2 
    * CASSSNAGAVGYTF
  * Alpha Chain
    * TRAV26-1
    * TRAJ27
    * CIVRVSYNTNAGKSTF
```{r, eval=FALSE}
# Run in terminal locally after installation
gui_stitchr
```


### -----------------------------

# BONUS - Finding Canditate Antigens

# 1. Load input dataframes

### a. load inputs and results
```{r}
# autoantigen screen
df_pep <- read.csv(counts_pep_fh,stringsAsFactors = F,row.names = 1)
results_pep <- read.csv(results_pep_fh,stringsAsFactors = F)

# viral antigen screen
df_vir <- read.csv(counts_vir_fh,stringsAsFactors = F,row.names = 1)
results_vir <- read.csv(results_vir_fh,stringsAsFactors = F)
```

### b. apply log10 to normalized counts
```{r}
df_pep_log <- log10(df_pep + 1)
df_vir_log <- log10(df_vir + 1)
```

# 2. Identify highly enriched autoantigens targeted by mAbs

### a. mAb1 filter for highly enriched peptides

* Use this code for mAb1 and mAb2. To plot the top enriched peptides for mAb2, change **samples_enriched_mAb1** to **samples_enriched_mAb2**.
```{r}
TOP_PEPTIDES <- 50

# filter for peptides only enriched in one target mAb
top_hits_pep_mab <- results_pep[results_pep$samples_enriched_mAb1 > 0,]
top_hits_pep_mab <- top_hits_pep_mab[top_hits_pep_mab$sum_expression_controls < 10,]

# Create a ratio of summed expression of mAb relative to the controls
top_hits_pep_mab$ratio <- (top_hits_pep_mab$sum_expression_mAb1 + 1) / (top_hits_pep_mab$sum_expression_controls + 1)

# Sort by this ratio
top_hits_pep_mab            <- top_hits_pep_mab[order(top_hits_pep_mab$ratio,decreasing = T),]
top_pep_mab_peptides        <- top_hits_pep_mab$peptide_id[1:TOP_PEPTIDES]
names(top_pep_mab_peptides) <- tstrsplit(top_pep_mab_peptides,"_")[[1]]
top_ordered_genes           <- unique(names(top_pep_mab_peptides))
```

### b. prepare plot dataframe plus cluster rows and columns
```{r}
df_plot_pep <- df_pep_log[row.names(df_pep_log) %in% top_pep_mab_peptides,]

# list mAb and Control samples
all_cols <- names(df_plot_pep)
mab1_samples    <- all_cols[grep("mAb1",all_cols)]
mab2_samples    <- all_cols[grep("mAb2",all_cols)]
control_samples <- all_cols[grep("Control",all_cols)]

# Cluster sample columns, specifically the controls
control_samples_ordered <- getGroupSampleOrderHeatmap(df_plot_pep,select_samples = control_samples,select_peptides = row.names(df_plot_pep))

df_plot_pep <- df_plot_pep[,c(mab1_samples,mab2_samples,control_samples_ordered)]

# Cluster peptide rows and keep peptides from the same gene together
pep_order <- c()
for (gene in top_ordered_genes) {
  gene_peps <- top_pep_mab_peptides[names(top_pep_mab_peptides) %in% gene]
  
  if(length(gene_peps) > 1){
    pep_order_gene <- getGroupPeptideOrderHeatmap(df_plot_pep,select_samples = names(df_plot_pep),select_peptides = gene_peps)
  } else{
    pep_order_gene <- gene_peps
  }
  pep_order <- c(pep_order,pep_order_gene)
}

df_plot_pep <- df_plot_pep[pep_order,]
```


### c. setup heatmap parameters
```{r}
# Column annotations
col_annot            <- as.data.frame(names(df_plot_pep))
names(col_annot)     <- "type"
row.names(col_annot) <- col_annot$type
col_annot$type       <- str_replace_all(col_annot$type,"MS_patient1_|MS_patient2_","")
col_annot$type       <- tstrsplit(col_annot$type,"_")[[1]]

colcolor        <- c("purple","darkred","darkgreen")
names(colcolor) <- c("mAb1","mAb2","Control")
ANNOT_COLORS    <- list(type = colcolor)

# Custom color scheme with a break list from -1 to 4
break_1 <- seq(-1,1,0.04)  # first 51 breaks
break_2 <- seq(1,4,0.06)  # last 50 breaks
break_2 <- break_2[3:length(break_2)]
break_list <- c(break_1,break_2)

color_pal  <- bluered(100)
```


### d. plot heatmap - top candidate autoantigens
```{r,fig.height=6, fig.width=6}
hp_pep <- pheatmap(as.matrix(df_plot_pep),cluster_rows = F,cluster_cols = F,
                show_rownames = T,show_colnames = T, gaps_col = c(2,4),
                annotation_col = col_annot,annotation_colors = ANNOT_COLORS,
                border_color = NA,color = color_pal,breaks = break_list,
                fontsize = 7)
hp_pep
```


# 3. Identify highly enriched viral antigens targeted by mAbs

### a. mAb1 filter for highly enriched peptides
```{r}
# filter for peptides only enriched in one target mAb
top_hits_vir_mab1 <- results_vir[results_vir$samples_enriched_mAb1 > 0,]
top_hits_vir_mab1 <- top_hits_vir_mab1[top_hits_vir_mab1$sum_expression_controls < 10,]

# Filter further for peptides enriched across replicates
top_hits_vir_mab1 <- top_hits_vir_mab1[top_hits_vir_mab1$samples_enriched_mAb1 == 2,]

# Only keep human viruses
top_hits_vir_mab1 <- top_hits_vir_mab1[top_hits_vir_mab1$human_virus_status %in% "yes",]
```

### b. mAb2 filter for highly enriched peptides
```{r}
# filter for peptides only enriched in one target mAb
top_hits_vir_mab2 <- results_vir[results_vir$samples_enriched_mAb2 > 0,]
top_hits_vir_mab2 <- top_hits_vir_mab2[top_hits_vir_mab2$sum_expression_controls < 10,]

# Filter further for peptides enriched across replicates
top_hits_vir_mab2 <- top_hits_vir_mab2[top_hits_vir_mab2$samples_enriched_mAb2 == 2,]

# Only keep human viruses
top_hits_vir_mab2 <- top_hits_vir_mab2[top_hits_vir_mab2$human_virus_status %in% "yes",]
```


### c. prepare plot dataframe plus cluster rows and columns
```{r}
top_vir_mab_peptides <- unique(c(top_hits_vir_mab1$peptide,top_hits_vir_mab2$peptide))

df_plot_vir <- df_vir_log[row.names(df_vir_log) %in% top_vir_mab_peptides,]

# list mAb and Control samples
all_cols <- names(df_plot_vir)
mab1_samples     <- all_cols[grep("mAb1",all_cols)]
mab2_samples     <- all_cols[grep("mAb2",all_cols)]
control_samples <- all_cols[grep("Control",all_cols)]

# Cluster sample columns, specifically the controls
control_samples_ordered <- getGroupSampleOrderHeatmap(df_plot_vir,select_samples = control_samples,select_peptides = row.names(df_plot_vir))

df_plot_vir <- df_plot_vir[,c(mab1_samples,mab2_samples,control_samples_ordered)]


# Cluster peptide rows and keep peptides from the same viral family together
metadata_vir <- results_vir[results_vir$peptide %in% top_vir_mab_peptides,]
metadata_vir <- metadata_vir[,c("peptide","family","species","sequence")]
row.names(metadata_vir) <- metadata_vir$peptide

row_annot     <- clusterRowsByViralFamily(df_plot_vir,metadata_vir)
pep_order_vir <- row.names(row_annot)

df_plot_vir <- df_plot_vir[pep_order_vir,]

# order the peptide metadata the same way as the clustered plot dataframe
metadata_vir <- metadata_vir[pep_order_vir,]
```


### d. setup heatmap parameters
```{r}
# Column annotations
col_annot            <- as.data.frame(names(df_plot_vir))
names(col_annot)     <- "type"
row.names(col_annot) <- col_annot$type
col_annot$type       <- str_replace_all(col_annot$type,"MS_patient1_|MS_patient2_","")
col_annot$type       <- tstrsplit(col_annot$type,"_")[[1]]

colcolor        <- c("purple","darkred","darkgreen")
names(colcolor) <- c("mAb1","mAb2","Control")

# Row annotation colors
uni_fam  <- unique(metadata_vir$family)
rowcolor <- brewer.pal(n = 8, name = "Dark2")
rowcolor <- rowcolor[1:length(uni_fam)]
names(rowcolor) <- uni_fam

ANNOT_COLORS_VIR    <- list(type = colcolor,family=rowcolor)

# Custom color scheme with a break list from -1 to 4
break_1 <- seq(-1,1,0.04)  # first 51 breaks
break_2 <- seq(1,4,0.06)  # last 50 breaks
break_2 <- break_2[3:length(break_2)]
break_list <- c(break_1,break_2)

color_pal  <- bluered(100)
```


### e. plot heatmap - top candidate viral antigens

* View **metadata_vir** while looking at this heatmap to get more info on each peptide.
```{r,fig.height=4, fig.width=12}
hp_vir <- pheatmap(as.matrix(df_plot_vir),cluster_rows = F,cluster_cols = F,
                show_rownames = T,show_colnames = T, gaps_col = c(2,4),
                annotation_col = col_annot,annotation_row = row_annot,
                annotation_colors = ANNOT_COLORS_VIR,
                border_color = NA,color = color_pal,breaks = break_list,
                fontsize = 5)
hp_vir
```



